answers/src/main/java/com/coding/solutions/basic/SinglyLinkedList.java

```diff
@@ -0,0 +1,158 @@
+         */
+        if(head == null){
+            //since there is only one element, both head and
+            //tail points to the same object.
+            head = nd;
+            tail = nd;
+        } else {
+            //set current tail next link to new node
+            tail.setNext(nd);
+            //set tail as newly created node
+            tail = nd;
+        }
+    }
+
+    public void addAfter(T element, T after){
+
+        SinglyLinkedListNode<T> tmp = head;
+        SinglyLinkedListNode<T> refNode = null;
+        System.out.println("Traversing to all nodes..");
+        /**
+         * Traverse till given element
+         */
+        while(true){
+            if(tmp == null){
+                break;
+            }
+            if(tmp.compareTo(after) == 0){
+                //found the target node, add after this node
+                refNode = tmp;
+                break;
+            }
+            tmp = tmp.getNext();
+        }
+        if(refNode != null){
+            //add element after the target node
+            SinglyLinkedListNode<T> nd = new SinglyLinkedListNode<T>();
+            nd.setValue(element);
+            nd.setNext(tmp.getNext());
+            if(tmp == tail){
+                tail = nd;
+            }
+            tmp.setNext(nd);
+
+        } else {
+            System.out.println("Unable to find the given element...");
+        }
+    }
+
+    public void deleteFront(){
+
+        if(head == null){
+            System.out.println("Underflow...");
+        }
+        SinglyLinkedListNode<T> tmp = head;
+        head = tmp.getNext();
+        if(head == null){
+            tail = null;
+        }
+        System.out.println("Deleted: "+tmp.getValue());
+    }
+
+    public void deleteAfter(T after){
+
+        SinglyLinkedListNode<T> tmp = head;
+        SinglyLinkedListNode<T> refNode = null;
+        System.out.println("Traversing to all nodes..");
+        /**
+         * Traverse till given element
+         */
+        while(true){
+            if(tmp == null){
+                break;
+            }
+            if(tmp.compareTo(after) == 0){
+                //found the target node, add after this node
+                refNode = tmp;
+                break;
+            }
+            tmp = tmp.getNext();
+        }
+        if(refNode != null){
+            tmp = refNode.getNext();
+            refNode.setNext(tmp.getNext());
+            if(refNode.getNext() == null){
+                tail = refNode;
+            }
+            System.out.println("Deleted: "+tmp.getValue());
+        } else {
+            System.out.println("Unable to find the given element...");
+        }
+    }
+
+    public void traverse(){
+
+        SinglyLinkedListNode<T> tmp = head;
+        while(true){
+            if(tmp == null){
+                break;
+            }
+            System.out.println(tmp.getValue());
+            tmp = tmp.getNext();
+        }
+    }
+
+    public void reverse() {
```

Comment:
What's a good way to save the elements of the sublist that's not reversed ? The reverse method that I implemented discards the elements before the reverse start position.

Should I create a new object for the subList from node position N ?

```java
 public void reverse(int start) {
        if (this == null || this.head == null) {
            return;
        }

        SinglyLinkedListNode prev = null;
        SinglyLinkedListNode current = head;

        int index = 0;
        while (index < start) {
            current = current.getNext();
            index++;
        }

        SinglyLinkedListNode next = ( current != null) ? current.getNext() : null;
        SinglyLinkedListNode nextToNext = (next != null) ? next.getNext() : null;

        while ( current != null ) {
//            System.out.println(print(prev) + " -> " + print(current) + " -> " + print(next) + " -> " + print(nextToNext));
            current.setNext(prev);
            prev = current;
            current = next;
            next = nextToNext;
            if (nextToNext != null) nextToNext = nextToNext.getNext();
        }
        head = prev;
    }
```java